### PR TITLE
Issue #39 - Add SNS Support

### PR DESCRIPTION
**Pull request by toadkicker**
This is pretty basic, untested, etc. I'm posting this to get more feedback on the output, and I have a limited testing environment for chat clients. Expect more commits as I wouldn't call this a complete PR just yet.

### CODE REVIEW

**Comment by yoheimuta** [comment]
Thanks!
I'm going to check it.

**Comment by toadkicker** [comment]
@yoheimuta So far I have the list functions working which completes the story. However I added publish and I think that should be a part of the PR before its acceptable. I should finish testing it today.

**Comment by yoheimuta** [comment]
> I'm posting this to get more feedback on the output, and I have a limited testing environment for chat clients
> I'm going to check it.

I tried this pr in my environment. It sounds great.
I'm waiting for your work done, thanks !

**Comment by toadkicker** [comment]
@yoheimuta wrapping this PR up. Mind testing it and if all is good :shipit:

**Comment by yoheimuta** [comment]
I tested it, and it sounds good !

I got the warning message `WARNING Expected /path/to/hubot-aws/scripts/sns/sns-message-verify to assign a function to module.exports, got object` .
Could you find and solve this message ?

**Comment by toadkicker** [comment]
@yoheimuta fixed the warning message you mentioned. Good catch!

**Comment by yoheimuta** [comment]
Thanks, I checked it.

But I still saw the same message...
Are you sure this message is gone ?

Call a [robot.loadFile](https://github.com/yoheimuta/hubot-aws/blob/7ede399b0ac2ba77a81c9ec194ddc841f7cd51b0/index.coffee#L10) to sns-message-verify outputs this message.

So I tried avoiding this call and confirmed the message is gone.
Of course, I also confirmed that it works to receive topic messages properly.

``` diff
/m/d/j/h/g/h/s/sns ❯❯❯ gws
R  sns-message-verify.coffee -> lib/sns-message-verify.coffee
M  sns.coffee
/m/d/j/h/g/h/s/sns ❯❯❯ gwd --cached
diff --git a/scripts/sns/sns-message-verify.coffee b/scripts/sns/lib/sns-message-verify.coffee
similarity index 100%
rename from scripts/sns/sns-message-verify.coffee
rename to scripts/sns/lib/sns-message-verify.coffee
diff --git a/scripts/sns/sns.coffee b/scripts/sns/sns.coffee
index 0e380cb..6fca371 100644
--- a/scripts/sns/sns.coffee
+++ b/scripts/sns/sns.coffee
@@ -20,7 +20,7 @@
 #   mdouglass

 { inspect } = require 'util'
-{ verifySignature } = require './sns-message-verify'
+{ verifySignature } = require './lib/sns-message-verify'

 Options =
   url:     process.env.HUBOT_SNS_URL or '/hubot/sns'
```

How about this diff ?

**Comment by toadkicker** [comment]
Oh that makes sense, it is being loaded twice then.

**Comment by toadkicker** [comment]
@yoheimuta I changed the test to exclude the support folder from loading. The loop over the folder names produces a 'file' = 'support' which is why it fails.

**Comment by yoheimuta** [comment]
Thanks! I'm going to review it.

File: scripts/sns/sns.coffee [hunk 1]
```diff
@@ -86,7 +86,12 @@ class SNS
     @robot.emit 'sns:notification', message
     @robot.emit 'sns:notification:' + message.topic, message
 
+  channelID: (name) ->
+    temp = name.toLowerCase().replace(/\s/g, "_")
+    "#{process.env.HUBOT_HIPCHAT_JID.split("_")[0]}_#{temp}@conf.hipchat.com"
```
Comment (yoheimuta) [inline scripts/sns/sns.coffee]: [MEMO] This feature is for hipchat users.
If HUBOT_HIPCHAT_JID is not set, an error happens.

```
[Tue Aug 09 2016 12:18:13 GMT+0000 (UTC)] ERROR TypeError: Cannot call method 'split' of undefined
  at SNS.channelID (/tmp/hubot-aws/scripts/sns/sns.coffee:91:36, <js>:92:50)
  at EventEmitter.<anonymous> (/tmp/hubot-aws/scripts/sns/sns.coffee:96:5, <js>:103:36)
  at EventEmitter.emit (events.js:95:17)
  at Robot.emit (/tmp/mtburn_hubot_dev/node_modules/hubot/src/robot.coffee:583:18, <js>:459:41)
  at SNS.notify (/tmp/hubot-aws/scripts/sns/sns.coffee:86:5, <js>:85:18)
  at SNS.process (/tmp/hubot-aws/scripts/sns/sns.coffee:62:7, <js>:55:21)
  at /tmp/hubot-aws/scripts/sns/sns.coffee:46:11, <js>:35:26
  at /tmp/hubot-aws/scripts/sns/support/sns_message_verify.coffee:57:12, <js>:77:14
  at IncomingMessage.<anonymous> (/tmp/hubot-aws/scripts/sns/support/sns_message_verify.coffee:21:14, <js>:29:16)
  at IncomingMessage.EventEmitter.emit (events.js:117:20)
  at _stream_readable.js:920:16
  at process._tickCallback (node.js:415:13)
```

**Comment by yoheimuta** [comment]
I wrote a comment https://github.com/yoheimuta/hubot-aws/pull/47/commits/4440c3e6948437f450b9d066909601ddc4f85577#r74047075.

I'm ok if you are also ok that a feature of receiving a message is for hipchat users.
Can you check it to be sure ?

File: scripts/sns/sns.coffee (hunk 1 above)
Comment (toadkicker) [inline scripts/sns/sns.coffee]: We probably don't want hipchat specific code here. The point of this code was to make it easier for the sns message to be composed. Hipchat makes people dig to find the JID of a room.

Comment (toadkicker) [inline scripts/sns/sns.coffee]: I'll figure out a way to make this more generic

**Comment by toadkicker** [comment]
Yeah I realized how specific to my implementation that actually is. I don't really think it belongs here. Instead I wrote documentation for how to hook into the sns:notification event so developers can wire up whatever they want to that.

**Comment by yoheimuta** [comment]
I checked your commits.
I'm ok to merge this pr and release a new version.

Are you ok ?

**Comment by toadkicker** [comment]
@yoheimuta (shipit)

**Comment by yoheimuta** [comment]
I released v0.5.0. Thanks.
https://github.com/yoheimuta/hubot-aws/blob/master/CHANGELOG.md#v050-2016-08-19
https://www.npmjs.com/package/hubot-aws

**Pull request merged**